Feature/m0078 ddg 308 tutorial debug ui freeze (#54)
MicroEJ/docs: Tutorials/tutorialDebugUIFreeze.rst

Debug a “UI Freeze”
========================

When an application User Interface freezes and becomes unresponsive, in most cases, one of the following conditions applies: 

- An unrecoverable system failure occurred, like a HardFault, and the RTOS tasks are not scheduled anymore. 
- The RTOS task that runs the MicroEJ runtime is never given CPU time (suspended or blocked). 
- The RTOS task that runs the MicroEJ runtime is executing never-ending native code (infinite loop in native implementation for example). 
- A Java method is executing a long running operation in the MicroUI Display Pump thread. 
- The application code is unable to receive or process user input events.

This tutorial explains how to instrument the code in order to locate the issue when the UI freeze occurs.

> Bottom-Up approach

Reviewer: added.

------------------

Let’s start at low level by figuring out if the RTOS is scheduling the

> tasks correctly. Make one of the RTOS task act like a heart beat: create

Reviewer 2: I think we could also advise the reader to use the debugging features of their C IDE for this part.
Then, as a last resort (i.e. if it's not possible to run the application in debug mode), they can add a new task as described here.

Reviewer: Added.  I also added that the tutorial assumes the freeze only occurs on device and not on simulator.

> a dedicated task and make it report in some way at a regular pace (print

Reviewer 2: What should be this task's priority? Probably lower or equal than the one dedicated to run the VM?

Reviewer: I think it should be higher but I think ~~both~~ all three are worth the test.

- If the RTOS task is of higher priority, we are ensuring that the RTOS scheduler is still scheduling.
- If the RTOS task is of equal priority, we are ensuring that the RTOS scheduler should (theoretically) schedule the VM
- If the RTOS task is of lower priority, we are ensuring that the RTOS scheduler can schedule lower priority tasks. 

WDYT?

Reviewer: I added a paragraph and a diagram about the RTOS priorities. WDYT?

If the diagram is OK I'll redo it with MicroEJ graphic identity.

relying on any native RTOS capabilities. Therefore, the whole Java world
runs in one single RTOS task. Read more about this architecture in the
`Platform Developer
Guide <https://docs.microej.com/en/latest/PlatformDeveloperGuide/coreEngine.html>`__.
A quick way to check if the Java threads are scheduled correctly is, here again, to
make one of the threads print a heart beat message. Copy/paste the
following snippet in the ``main()`` method of the application:

::

   TimerTask task = new TimerTask() {

       @Override
       public void run() {
           System.out.println("Alive");
       }
   };
   Timer timer = new Timer();
   timer.schedule(task, 10_000, 10_000);

This code creates a new Java thread that will print the message “Alive”
on the standard output every 10 seconds. If the “Alive” printouts stop
when the UI freeze occurs (assuming no one cancelled the ``Timer``),
then it means that the MicroEJ Runtime stopped scheduling the Java
threads. Few suggestions:

-  The RTOS task that runs the MicroEJ runtime might be suspended or
   blocked. Check if some API call is suspending the task or if a shared
   resource could be blocking it.

> -  When a Java native method executes, it executes its C counterpart

Reviewer: fixed.

      ``validateContent()`` and ``setBoundsContent()``
   -  ``handleEvent()``
   -  ``callSerially()``: code wrapped in such calls will execute in the
      UI thread

-  The UI thread has terminated.

As a general rule, avoid running long operations in the UI thread,
follow the general pattern and use a dedicated thread/executor instead:
(QUESTION: use a sequence diagram to be more explicit?)

::

   ExecutorService executorService = ServiceLoaderFactory.getServiceLoader().getService(ExecutorService.class, SingleThreadExecutor.class);
   executorService.execute(new Runnable() {

       @Override
       public void run() {
       
           // (... long non-UI operation ...)
           
           // optional: update the UI upon completion
           Display.getDefaultDisplay().callSerially(new Runnable() {
                       
           @Override
           public void run() {
                       // update display code (will be executed in UI thread)
           }
       });
       }
   });

Another case that is worth looking at is whether the application is
processing user input events like it should. The UI may look “frozen”

> only because it don’t react to input events. Replace the desktop

Reviewer 3: ```suggestion
only because it doesn't react to input events. Replace the desktop
```

Reviewer: fixed.

Debug a "UI Freeze"
===================

When an application User Interface freezes and becomes unresponsive, in most cases, one of the following conditions applies: 

- An unrecoverable system failure occurred, like a HardFault, and the RTOS tasks are not scheduled anymore. 
- The RTOS task that runs the MicroEJ runtime is never given CPU time (suspended or blocked). 
- The RTOS task that runs the MicroEJ runtime is executing never-ending native code (infinite loop in native implementation for example). 
- A Java method is executing a long running operation in the MicroUI thread (also called Display Pump thread). 
- The application code is unable to receive or process user input events.

This tutorial explains how to instrument the code in order to locate the issue when the UI freeze occurs.

Bottom-Up approach
------------------

> Check RTOS scheduler liveness

Reviewer 4: ```suggestion
Check RTOS Scheduler Liveness
```

tasks correctly.

Make one of the RTOS task act like a heart beat: create a dedicated
task and make it report in some way at a regular pace (print a message
on standard output, blink a LED, etc.).

If the heart beat is still running when the UI freeze occurs, we can
go a step further and check whether the MicroEJ runtime is still
scheduling Java threads or not.

Check Java Scheduler Liveness
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As a reminder, the architecture of the MicroEJ runtime is called green
thread architecture, it defines a multi-threaded environment without
relying on any native RTOS capabilities. Therefore, the whole Java world
runs in one single RTOS task. Read more about this architecture in the
`Platform Developer
Guide <https://docs.microej.com/en/latest/PlatformDeveloperGuide/coreEngine.html>`__.
A quick way to check if the Java threads are scheduled correctly is, here again, to
make one of the threads print a heart beat message. Copy/paste the
following snippet in the ``main()`` method of the application:

.. code-block:: java

   TimerTask task = new TimerTask() {

       @Override
       public void run() {
           System.out.println("Alive");
       }
   };

>    Timer timer = new Timer();

Reviewer 4: If the number of threads available in the platform is fitted to the actual number of threads running, the user have to increment this value in the launch configuration. And it may not be possible if the memory is already full.

Reviewer 4: Maybe propose to reuse another thread to add this trace (an existing timer or a thread where these kind of periodic traces are easy to put)

Reviewer: Good point.

Reviewer: I'm considering adding a new section named `Implementation details` or `Implementation consideration` to put this information, and the other about blinking LED.

My idea is to reduce overloading the debugging process with too many implementation details.
What do you think?

Reviewer 4: Yep, I agree! With a link (in the form `(1)` for example).

Reviewer: I added a `Implementation details` section. WDYT?

Reviewer 4: Perhaps add a link to this section when the user has to create a new thread/timer.

Reviewer: added

   timer.schedule(task, 10_000, 10_000);

> This code creates a new Java thread that will print the message ``Alive``

Reviewer 4: If the standard output cannot be read, suggest to use a LED or another way as in the RTOS section.

Reviewer 4: It may also be a trace system such as system view.

Reviewer: I added a note to suggest blinking a LED.  I'm reluctant to mention systemview since we don't provide any doc for that (yet).

Author: There is no section about SystemView in the docs, but we already mention SystemView in the documentation and in other tutorials. imho, mentioning it could give the hint that this solution exists and that it has already been used in MicroEJ environments. https://docs.microej.com/en/latest/ApplicationDeveloperGuide/trace.html#trace-implementations

Reviewer: added

on the standard output every 10 seconds. If the ``Alive`` printouts stop
when the UI freeze occurs (assuming no one cancelled the ``Timer``),

> then it means that the MicroEJ Runtime stopped scheduling the Java

Reviewer 4: It could also means that the Timer is:
- canceled,
- stuck in the execution of a timer task.

threads.

Here are a few suggestions:

- The RTOS task that runs the MicroEJ runtime might be suspended or
  blocked. Check if some API call is suspending the task or if a
  shared resource could be blocking it.

- When a Java native method is called, it calls its C counterpart
  function in the RTOS task that runs the MicroEJ runtime. While the C
  function is running, no other Java methods can run : the Java world
  awaits for the C function to finish. As a consequence, if the C
  function never returns, no Java thread can ever run again. Spot any

>   suspect native functions and print every entry/exit to detect faulty

Reviewer 4: `trace` instead of `print`

Reviewer: Fixed. thanks.

In case this snippet prints ``TimerTask Alive`` but not ``UI alive`` when
the freeze occurs, then there are few options:

-  The application might be processing a long operation in the UI
   thread, for example:

   -  infinite/indeterminate loops
   -  network/database access
   -  heavy computations
   -  ``Thread.sleep()``/``Object.wait()``
   -  ``SNI_suspendCurrentJavaThread()`` in native call

   When doing so, any other UI-related operation will not be processed
   until completion, leading the display to being unresponsive. Any code
   that runs in the UI thread might be responsible. Look for code
   executed as a result of calls to:

   -  ``repaint()``: code in ``renderContent()``
   -  ``revalidate()``/``revalidateSubTree()``: code in
      ``validateContent()`` and ``setBoundsContent()``
   -  ``handleEvent()``
   -  ``callSerially()``: code wrapped in such calls will be executed
      in the UI thread

-  The UI thread has terminated.

As a general rule, avoid running long operations in the UI thread,
follow the general pattern and use a dedicated thread/executor instead:

.. (QUESTION: use a sequence diagram to be more explicit?)

.. code-block:: java

>    ExecutorService executorService = ServiceLoaderFactory.getServiceLoader().getService(ExecutorService.class, SingleThreadExecutor.class);

Reviewer 4: Are you sure that we have the ExecutorService implementation? I know we have an Executor one though.

Reviewer: SingleThreadExecutor implements ExecutorService (according to https://repository.microej.com/javadoc/microej_5.x/apis/ej/util/concurrent/SingleThreadExecutor.html)

Reviewer 4: Ok, I didn't remember that 👍

           Display.getDefaultDisplay().callSerially(new Runnable() {
                       
           @Override
           public void run() {
                       // update display code (will be executed in UI thread)
           }
       });
       }
   });

Check Input Events Processing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Another case that is worth looking at is whether the application is
processing user input events like it should. The UI may look "frozen"
only because it doesn't react to input events. Replace the desktop
instance with the one below to log all user inputs.

.. code-block:: java

   Desktop desktop = new Desktop() {

       @Override
       public EventHandler getController() {
           EventHandler controller = super.getController();
           return new EventHandler() {
           @Override
               public boolean handleEvent(int event) {
               System.out.println("Desktop.handleEvent() received event of type " + Event.getType(event));
               return controller.handleEvent(event);
               }
           };
       }
   };

LLMJVM_dump
-----------

In addition to the tips described above, the ``LLMJVM_dump`` function

> will give detailled information about the current state of Java

Reviewer 4: `detailed` instead of `detailled`.

Reviewer: Fixed, thanks.

           }
       });
       }
   });

Check Input Events Processing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Another case that is worth looking at is whether the application is
processing user input events like it should. The UI may look "frozen"
only because it doesn't react to input events. Replace the desktop
instance with the one below to log all user inputs.

.. code-block:: java

   Desktop desktop = new Desktop() {

       @Override
       public EventHandler getController() {
           EventHandler controller = super.getController();
           return new EventHandler() {
           @Override
               public boolean handleEvent(int event) {
               System.out.println("Desktop.handleEvent() received event of type " + Event.getType(event));
               return controller.handleEvent(event);
               }
           };
       }
   };

LLMJVM_dump
-----------

In addition to the tips described above, the ``LLMJVM_dump`` function
will give detailed information about the current state of Java
threads when the issue occurs.  See :ref:`this section<vm_dump>` to
learn more about ``LLMJVM_dump``.

Implementation Details
----------------------

> - The number of threads in the MicroEJ Application must to be

Reviewer 4: ```suggestion
- The number of threads in the MicroEJ Application must be
```

Reviewer: added

  sufficient to support the creation of additional threads when using

>   ``Timer`` and ``Thread``.

Reviewer 4: Do we have a section in the main documentation that explain how to configure the threads? It could be linked from here.

Reviewer 4: Such as `Category: Memory` in https://docs.microej.com/en/latest/ApplicationDeveloperGuide/applicationOptions.html#category-runtime

Reviewer: added

  blocked. Check if some API call is suspending the task or if a
  shared resource could be blocking it.

- When a Java native method is called, it calls its C counterpart
  function in the RTOS task that runs the MicroEJ runtime. While the C
  function is running, no other Java methods can run : the Java world
  awaits for the C function to finish. As a consequence, if the C
  function never returns, no Java thread can ever run again. Spot any
  suspect native functions and trace every entry/exit to detect faulty
  code.

Please refer to :ref:`implementation_details` if you encounter issues
to implement the heart beat.

Check UI Thread Liveness
~~~~~~~~~~~~~~~~~~~~~~~~

Now, what if the ``Alive`` heart beat runs while the UI is frozen?
Java threads are getting scheduled but the UI thread (also called
Display Pump thread), does not process display events.

Let's make the heart beat snippet above execute in the UI
thread. Simply wraps the ``System.out.println("Alive")`` with a
``callSerially``:

.. code-block:: java

   TimerTask task = new TimerTask() {

       @Override
       public void run() {
           System.out.println("TimerTask Alive");

>            Display.getDefaultDisplay().callSerially(new Runnable() {

Reviewer 2: This doesn't work anymore with MicroUI3.

```suggestion
           MicroUI.callSerially(new Runnable() {
```
